File: src/policy/threat-intel-guard.ts
import { readFileSync, existsSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { resolveThreatStatePath } from '../ai/ai-paths.js';
import { walkStringLeaves } from './arg-leaf-walker.js';
import type { CallContext, PolicyDecision } from './policy-types.js';
import { getMtxThreatPatterns } from '../utils/mtx-threat-intel-bridge.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const REPO_ROOT = join(__dirname, '..', '..');
const SIGNATURES_PATH = join(REPO_ROOT, 'config', 'threat-intel-signatures.json');

let cachedPatterns: RegExp[] | null = null;
let cachedAt = 0;
const CACHE_TTL_MS = 60_000;

function escapeRegexLiteral(text: string): string {
  return text.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function compilePattern(source: string): RegExp | null {
  const trimmed = source.trim();
  if (!trimmed) return null;
  try {
    return new RegExp(trimmed, 'i');
  } catch {
    return new RegExp(escapeRegexLiteral(trimmed), 'i');
  }
}

function loadBaselinePatterns(): RegExp[] {
  if (!existsSync(SIGNATURES_PATH)) return [];
  try {
    const data = JSON.parse(readFileSync(SIGNATURES_PATH, 'utf-8')) as { patterns?: string[] };
    return (data.patterns ?? [])
      .map((p) => compilePattern(p))
      .filter((p): p is RegExp => p !== null);
  } catch {
    return [];
  }
}

function loadDynamicPatterns(): RegExp[] {
  try {
    const path = resolveThreatStatePath();
    if (!existsSync(path)) return [];
    const data = JSON.parse(readFileSync(path, 'utf-8')) as {
      entries?: Array<{ signature?: string; description?: string; severity?: string }>;
      catalog?: Array<{ signature?: string; description?: string; severity?: string }>;
    };
    const entries = data.entries ?? data.catalog ?? [];
    const patterns: RegExp[] = [];
    for (const entry of entries) {
      if (entry.signature) {
        const compiled = compilePattern(entry.signature);
        if (compiled) patterns.push(compiled);
      }
      if (entry.description && (entry.severity === 'CRITICAL' || entry.severity === 'HIGH')) {
        const snippet = entry.description.slice(0, 120).trim();
        if (snippet.length >= 24) {
          const compiled = compilePattern(snippet);
          if (compiled) patterns.push(compiled);
        }
      }
    }
    return patterns;
  } catch {
    return [];
  }
}

function loadMtxHashPatterns(): RegExp[] {
  return getMtxThreatPatterns()
    .map((hash) => compilePattern(hash.slice(0, 64)))
    .filter((p): p is RegExp => p !== null);
}

function getPatterns(): RegExp[] {
  if (process.env.MASTYFF_AI_DISABLE_THREAT_INTEL_GUARD === 'true') return [];
  const now = Date.now();
  if (cachedPatterns && now - cachedAt < CACHE_TTL_MS) return cachedPatterns;
  cachedPatterns = [...loadBaselinePatterns(), ...loadDynamicPatterns(), ...loadMtxHashPatterns()];
  cachedAt = now;
  return cachedPatterns;
}

/** Reset cached patterns (tests). */
export function resetThreatIntelGuardCache(): void {
  cachedPatterns = null;
  cachedAt = 0;
}

/** Block tool calls whose arguments match live or baseline threat-intel signatures. */
export function evaluateThreatIntelGuard(ctx: CallContext): PolicyDecision | null {
  const patterns = getPatterns();
  if (patterns.length === 0) return null;

  const blob = walkStringLeaves(ctx.arguments ?? {})
    .map((leaf) => leaf.value)
    .join('\n');
  if (!blob) return null;

  for (const pattern of patterns) {
    if (pattern.test(blob)) {
      return {
        action: 'block',
        rule: 'threat-intel',
        reason: 'Threat intel signature matched in tool arguments',
      };
    }
  }
  return null;
}

File: adversarial-harness/python/policy_engine/threat_intel_guard.py
"""Threat intel guard — Python port of src/policy/threat-intel-guard.ts."""

from __future__ import annotations

import json
import os
import re
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parents[3]
SIGNATURES_PATH = REPO_ROOT / "config" / "threat-intel-signatures.json"

_cached_patterns: list[re.Pattern[str]] | None = None
_cached_at: float = 0.0
_CACHE_TTL_SEC = 60.0


def _compile_pattern(source: str) -> re.Pattern[str] | None:
    trimmed = source.strip()
    if not trimmed:
        return None
    try:
        return re.compile(trimmed, re.I)
    except re.error:
        escaped = re.escape(trimmed)
        return re.compile(escaped, re.I)


def _load_baseline_patterns() -> list[re.Pattern[str]]:
    if not SIGNATURES_PATH.is_file():
        return []
    try:
        data = json.loads(SIGNATURES_PATH.read_text(encoding="utf-8"))
        patterns: list[re.Pattern[str]] = []
        for raw in data.get("patterns") or []:
            compiled = _compile_pattern(str(raw))
            if compiled:
                patterns.append(compiled)
        return patterns
    except (OSError, json.JSONDecodeError):
        return []


def _load_dynamic_patterns() -> list[re.Pattern[str]]:
    state_path = os.environ.get("MASTYFF_AI_THREAT_STATE_PATH")
    if not state_path:
        state_path = str(REPO_ROOT / ".threat-state.json")
    path = Path(state_path)
    if not path.is_file():
        return []
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        entries = data.get("entries") or data.get("catalog") or []
        patterns: list[re.Pattern[str]] = []
        for entry in entries:
            signature = entry.get("signature")
            if signature:
                compiled = _compile_pattern(str(signature))
                if compiled:
                    patterns.append(compiled)
            desc = entry.get("description")
            severity = entry.get("severity")
            if desc and severity in ("CRITICAL", "HIGH"):
                snippet = str(desc)[:120].strip()
                if len(snippet) >= 24:
                    compiled = _compile_pattern(snippet)
                    if compiled:
                        patterns.append(compiled)
        return patterns
    except (OSError, json.JSONDecodeError):
        return []


def _get_patterns() -> list[re.Pattern[str]]:
    global _cached_patterns, _cached_at
    if os.environ.get("MASTYFF_AI_DISABLE_THREAT_INTEL_GUARD") == "true":
        return []
    import time

    now = time.time()
    if _cached_patterns is not None and now - _cached_at < _CACHE_TTL_SEC:
        return _cached_patterns
    _cached_patterns = _load_baseline_patterns() + _load_dynamic_patterns()
    _cached_at = now
    return _cached_patterns


def reset_threat_intel_guard_cache() -> None:
    global _cached_patterns, _cached_at
    _cached_patterns = None
    _cached_at = 0.0


def evaluate_threat_intel_guard(arguments: dict | None) -> tuple[str, str] | None:
    from .arg_walker import walk_string_leaves

    patterns = _get_patterns()
    if not patterns:
        return None
    blob = "\n".join(leaf.value for leaf in walk_string_leaves(arguments or {}))
    if not blob:
        return None
    for pattern in patterns:
        if pattern.search(blob):
            return ("threat-intel", "Threat intel signature matched in tool arguments")
    return None
